runtime/include/chpl-tuner.h

```diff
@@ -0,0 +1,104 @@
+ * limitations under the License.
+ */
+
+#ifndef _chpl_tuners_h_
+#define _chpl_tuners_h_
+
+#include <stdint.h>
+#include "chpltypes.h"
+
+extern void* chpl_tuner_default_session;
+
+// ===================================================================
+// Chapel runtime third-party performance tuner interface.
+//
+// Performance tuners search for optimal values of variables given a
+// valid value range.  This file declares the interface to communicate
+// with external performance tuners.
+//
+// Third-party tuners should implement chpl_tuner_session_*_hook()
+// functions.  These are called from the Chapel runtime tuning
+// framework.
+// ===================================================================
+
+//
+// Initialize a new tuning session.
+//
+// This function should return a unique identifier to be used with the
+// remaining functions declared in this file.
+//
+void* chpl_tuner_session_init_hook(void);
+
+//
+// Initialize a new variable within a tuning session.
+//
+void chpl_tuner_session_newvar_hook(void*    session,
+                                    c_string name,
+                                    _real64* val_ptr,
+                                    _real64  min,
+                                    _real64  max,
+                                    _real64  step);
+
+//
+// Begin generating values for all tuning variables declared thus far.
+//
+void chpl_tuner_session_start_hook(void* session);
+
+//
+// End the search for optimal values.
+//
+// This function is distinct from chpl_tuner_session_fini_hook() for
+// the "restart" case, when a session is stopped and started without
+// releasing associated resources.
+//
+void chpl_tuner_session_stop_hook(void* session);
+
+//
+// Release all resources associated with the specified tuning session.
+//
+void chpl_tuner_session_fini_hook(void* session);
+
+//
+// Report the performance of a loop.
+//
+void chpl_tuner_session_loop_hook(void*   session,
+                                  _real64 performance);
+
+// ===================================================================
+// Chapel runtime framework function declarations.
+//
+// Third-party tuners should not implement these functions.
+// ===================================================================
+
+void*   chpl_tuner_session_init(uint32_t  lineno,
+                                c_string  filename);
+void    chpl_tuner_session_fini(void*    session,
+                                uint32_t lineno,
+                                c_string filename);
+void    chpl_tuner_session_loop_span(void*    session,
+                                     uint32_t count);
+_real64 chpl_tuner_get(void*    session,
+                       _real64  min,
+                       _real64  max,
+                       _real64  step,
+                       _real64  init,
+                       uint32_t lineno,
+                       c_string filename);
```

Comment:
So, if we had the "Caller ID" function, could we not even have any new C code in runtime/ ?